File: app/src/main/java/me/ykrank/s1next/widget/glide/AvatarStreamFetcher.java
package me.ykrank.s1next.widget.glide;

import com.bumptech.glide.Priority;
import com.bumptech.glide.integration.okhttp3.OkHttpStreamFetcher;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.Key;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.util.ContentLengthInputStream;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.inject.Inject;

import me.ykrank.s1next.App;
import me.ykrank.s1next.data.pref.DownloadPreferencesManager;
import me.ykrank.s1next.widget.glide.model.AvatarUrl;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.internal.cache.CacheStrategy;

/**
 * Fetches an {@link java.io.InputStream} from {@link AvatarUrl} using the OkHttp library.
 * <p>
 * Forked from {@link OkHttpStreamFetcher}
 */

public class AvatarStreamFetcher implements DataFetcher<InputStream> {
    private final AvatarUrl url;
    private final Call.Factory client;
    private InputStream stream;
    private ResponseBody responseBody;
    private volatile Call call;

    @Inject
    DownloadPreferencesManager mDownloadPreferencesManager;
    @Inject
    AvatarUrlsCache avatarUrlsCache;

    public AvatarStreamFetcher(Call.Factory client, AvatarUrl url) {
        this.url = url;
        this.client = client;
        App.getAppComponent().inject(this);
    }

    @Override
    public void loadData(Priority priority, DataCallback<? super InputStream> callback) {
        if (!mDownloadPreferencesManager.isAvatarsDownload()) {
            callback.onDataReady(null);
            return;
        }
        String url_string = url.toStringUrl();
        //whether cached error url
        final Key avatarKey = OriginalKey.obtainAvatarKey(mDownloadPreferencesManager, url_string);
        if (avatarUrlsCache.has(avatarKey)) {
            // already have cached this not success avatar url
            callback.onDataReady(null);
            return;
        }

        Request.Builder requestBuilder = new Request.Builder().url(url_string);

        for (Map.Entry<String, String> headerEntry : url.getHeaders().entrySet()) {
            String key = headerEntry.getKey();
            requestBuilder.addHeader(key, headerEntry.getValue());
        }

        Request request = requestBuilder.build();
        call = client.newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@Nonnull Call call, @Nullable IOException e) {
                if (avatarKey != null) {
                    avatarUrlsCache.put(avatarKey);
                }
                callback.onLoadFailed(e);
            }

            @Override
            public void onResponse(@Nonnull Call call, @Nonnull Response response) throws IOException {
                responseBody = response.body();
                if (!response.isSuccessful()) {
                    // if (this this a avatar URL) && (this URL is cacheable)
                    if (avatarKey != null && CacheStrategy.isCacheable(response, request)) {
                        avatarUrlsCache.put(avatarKey);
                        callback.onDataReady(null);
                        return;
                    }
                    callback.onLoadFailed(new IOException("Request failed with code: " + response.code()));
                } else {
                    // if download success, and (this this a avatar URL) && (this URL is cacheable)
                    // remove from cache list
                    if (avatarKey != null && avatarUrlsCache.has(avatarKey)) {
                        avatarUrlsCache.remove(avatarKey);
                    }
                    if (responseBody != null) {
                        long contentLength = responseBody.contentLength();
                        stream = ContentLengthInputStream.obtain(responseBody.byteStream(), contentLength);
                    }

                    callback.onDataReady(stream);
                }
            }
        });
    }

    @Override
    public void cleanup() {
        try {
            if (stream != null) {
                stream.close();
            }
        } catch (IOException e) {
            // Ignored
        }
        if (responseBody != null) {
            responseBody.close();
        }
    }

    @Override
    public void cancel() {
        Call local = call;
        if (local != null) {
            local.cancel();
        }
    }

    @Override
    public Class<InputStream> getDataClass() {
        return InputStream.class;
    }

    @Override
    public DataSource getDataSource() {
        return DataSource.REMOTE;
    }
}

File: app/src/main/java/me/ykrank/s1next/widget/glide/AvatarStreamFetcher.kt
package me.ykrank.s1next.widget.glide

import com.bumptech.glide.Priority
import com.bumptech.glide.integration.okhttp3.OkHttpStreamFetcher
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.util.ContentLengthInputStream
import me.ykrank.s1next.App
import me.ykrank.s1next.data.pref.DownloadPreferencesManager
import me.ykrank.s1next.widget.glide.model.AvatarUrl
import me.ykrank.s1next.widget.image.ImageBiz
import okhttp3.*
import okhttp3.internal.cache.CacheStrategy
import java.io.IOException
import java.io.InputStream
import javax.inject.Inject

/**
 * Fetches an [java.io.InputStream] from [AvatarUrl] using the OkHttp library.
 *
 *
 * Forked from [OkHttpStreamFetcher]
 */

class AvatarStreamFetcher(
    private val client: Call.Factory,
    private val url: AvatarUrl,
) : DataFetcher<InputStream> {
    private var stream: InputStream? = null
    private var responseBody: ResponseBody? = null
    @Volatile
    private var call: Call? = null

    @Inject
    internal lateinit var mDownloadPreferencesManager: DownloadPreferencesManager
    @Inject
    internal lateinit var avatarFailUrlsCache: AvatarFailUrlsCache

    private val imageBiz by lazy {
        ImageBiz(mDownloadPreferencesManager)
    }

    init {
        App.appComponent.inject(this)
    }

    override fun loadData(priority: Priority, callback: DataFetcher.DataCallback<in InputStream>) {
        if (!mDownloadPreferencesManager.isAvatarsDownload && !url.forcePass) {
            callback.onDataReady(null)
            return
        }
        val urlString = url.toStringUrl()
        //whether cached error url
        val avatarKey = imageBiz.avatarCacheKey(urlString)
        if (!url.forcePass && avatarFailUrlsCache.has(avatarKey)) {
            // already have cached this not success avatar url
            callback.onDataReady(null)
            return
        }

        val requestBuilder = Request.Builder().url(urlString)

        for ((key, value) in url.headers) {
            requestBuilder.addHeader(key, value)
        }

        val request = requestBuilder.build()
        call = client.newCall(request)
        call?.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (avatarKey != null) {
                    avatarFailUrlsCache.put(avatarKey)
                }
                callback.onLoadFailed(e)
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                responseBody = response.body
                if (!response.isSuccessful) {
                    // if (this this a avatar URL) && (this URL is cacheable)
                    if (avatarKey != null && CacheStrategy.isCacheable(response, request)) {
                        avatarFailUrlsCache.put(avatarKey)
                        callback.onDataReady(null)
                        return
                    }
                    callback.onLoadFailed(IOException("Request failed with code: " + response.code))
                } else {
                    // if download success, and (this this a avatar URL) && (this URL is cacheable)
                    // remove from cache list
                    if (avatarKey != null && avatarFailUrlsCache.has(avatarKey)) {
                        avatarFailUrlsCache.remove(avatarKey)
                    }
                    responseBody?.let {
                        val contentLength = it.contentLength()
                        stream = ContentLengthInputStream.obtain(it.byteStream(), contentLength)
                    }
                    callback.onDataReady(stream)
                }
            }
        })
    }

    override fun cleanup() {
        try {
            stream?.close()
        } catch (e: IOException) {
            // Ignored
        }

        responseBody?.close()
    }

    override fun cancel() {
        call?.cancel()
    }

    override fun getDataClass(): Class<InputStream> {
        return InputStream::class.java
    }

    override fun getDataSource(): DataSource {
        return DataSource.REMOTE
    }
}
